pkg/broker/handler/pool/fanout/pool.go

```diff
@@ -40,56 +39,59 @@ import (
 // It will also stop/delete the handler if the corresponding broker is deleted
 // in the config.
 type SyncPool struct {
-	options *pool.Options
-	targets config.ReadonlyTargets
-	pool    sync.Map
+	options        *pool.Options
+	targetsWatcher config.TargetsWatcher
+	pool           map[string]*handler.Handler
 }
 
 // StartSyncPool starts the sync pool.
-func StartSyncPool(ctx context.Context, targets config.ReadonlyTargets, opts ...pool.Option) (*SyncPool, error) {
+func StartSyncPool(ctx context.Context, targetsWatcher config.TargetsWatcher, opts ...pool.Option) (*SyncPool, error) {
 	p := &SyncPool{
-		targets: targets,
-		options: pool.NewOptions(opts...),
-	}
-	if err := p.syncOnce(ctx); err != nil {
-		return nil, err
-	}
-	if p.options.SyncSignal != nil {
-		go p.watch(ctx)
+		targetsWatcher: targetsWatcher,
+		options:        pool.NewOptions(opts...),
+		pool:           make(map[string]*handler.Handler),
 	}
+	go p.watch(ctx)
 	return p, nil
 }
 
 func (p *SyncPool) watch(ctx context.Context) {
+	targets := p.targetsWatcher.Targets()
 	for {
+		if err := p.syncTargets(ctx, targets.Config); err != nil {
+			logging.FromContext(ctx).Error("failed to sync handlers pool on watch signal", zap.Error(err))
+		}
 		select {
 		case <-ctx.Done():
 			return
-		case <-p.options.SyncSignal:
-			if err := p.syncOnce(ctx); err != nil {
-				logging.FromContext(ctx).Error("failed to sync handlers pool on watch signal", zap.Error(err))
-			}
+		case <-targets.Updated:
+			targets = p.targetsWatcher.Targets()
```

Comment:
It's slightly different, this signal only indicates that the current config is out of date and it can broadcast to any number of listeners.

The old signal channel required every signal to pulled downstream or the watcher would block and it also did not support broadcast. It also did not tie the signal to the config, so if the config was updated multiple times before a listener had a chance to sync the listener would still receive multiple signals and sync more than necessary.

With TargetsWatcher, a listener receives the latest config along with an update signal tied to that specific config value. The listener can use this signal to determine when that config value should be updated, but is not required to receive from that channel and the watcher will not be blocked.

Reply:
I guess I'm fine with `TargetsWatcher` if it provides improvements like you mentioned here. But I'd like the syncpool being the only one handles update signal. Not handler, not processors.